### DISCUSSION TITLE
Make the boilerplate more framework agnostic

### DISCUSSION THREAD

**Discussion by vplasencia** [semaphore-protocol/semaphore#685] [General]
It would be nice to create the boilerplate in a way that it's as framework-agnostic as possible. Right now it's hard to reproduce the boilerplate if you don't know much about React. There are some developers using other frameworks like `Vue` or `Angular`. It may be good for hackers at hackathons to easily reproduce the boilerplate in case they want to use another framework. We could replace React hooks with functions. What do you think?

**Comment by cedoor** [comment]
I agree with the need to make the boilerplate more framework-agnostic 👍🏽 How can we do it though?

The approach we have taken so far has been to use some of the most popular technologies in blockchain/zk projects: Nextjs and React. Looks like a good compromise between the use of modern technology and usability considering the average dev background, but it still doesn't support *all* devs. In particular, it doesn't support devs without any knowledge of React and NextJS, as you pointed out.

We could replace the hooks with functions that can be used with the other frameworks as well, but how can we build components that can be easily re-used in both Vue, Angular, and React?

We might consider using Web Components directly, which are framework-agnostic and supported by modern browsers.

Another alternative that comes to my mind could be to create a CLI template for each of the main front-end frameworks/libraries.

**Comment by vplasencia** [comment]
To create reusable code, we could use TypeScript code instead of hooks and context (which are React concepts). We could use just components and TypeScript code. Most JS frameworks support TS and the concept of components. 

For example, this project is using Semaphore with Next.js without hooks and context: https://github.com/bandada-infra/bandada-semaphore-demo

Another interesting point is the CSS library (Chakra UI) which is attached to React. We could consider using a framework like [`tailwindcss`](https://tailwindcss.com/) since it's easier to migrate an app because it uses HTML and is compatible with many JS frameworks. We could also keep Chakra here since the CLI templates use CSS and devs can use the CSS framework/library they prefer.

**Reply by cedoor** [reply]
`tailwindcss` is certainly more portable. I agree that boilerplate can actually continue to use Chakra.

I think I tend to prefer the idea of leaving the boilerplate as an example of an application built on top of Semaphore with popular JS technologies and adding more templates in the CLI, which seems like the right tool to cover the technologies we see being requested by the community.

**Reply by vplasencia** [reply]
I think it's fine if we update the boilerplate (without hooks and context) to make it easier to migrate to other JS frameworks. Then the boilerplate and templates keep the same structure and we can add in the documentation the main files that devs could copy/paste to migrate to other JS frameworks. What do you think?

**Reply by cedoor** [reply]
I've thought more about this and I'm not sure. It would not be a good example of how an application should be developed with React though. It would become a hybrid that you usually never see in an application of that type, with NextJS and React.

I think I still prefer the template as a place to offer alternatives to the other frameworks.

**Reply by vplasencia** [reply]
I mean, we will still use hooks but just when it's necessary. We wouldn't use context: https://github.com/semaphore-protocol/boilerplate/tree/main/apps/web-app/src/context. Also, this hook would be a Typescript file that you can use whenever you need: https://github.com/semaphore-protocol/boilerplate/blob/main/apps/web-app/src/hooks/useSemaphore.ts. The rest would be the same.

Then when devs want to use another framework, they will only copy/paste the equivalent of the Semaphore hook that will be a Typescript file without changing anything. I think it's a good way to make the boilerplate easier to understand for developers. 

I don't know if it makes sense to maintain many templates for different JS frameworks since there are a lot of frameworks and the code if we use a Typescript file with functions for that Semaphore hook would be the same. Also, it would be nice to have the same architecture for boilerplate and templates so that they all are easier to understand. 

I think that boilerplate and templates are a bit difficult to understand even if you know React and the idea is to make them easier to understand and migrate to other JS frameworks.

**Reply by cedoor** [reply]
Still not sure if I agree. When I see a React/NextJS app I expect it to be a React app, with React features (contexts and hooks). Mixing React with syntax devs usually use on non-react apps would make it more confusing IMO.

Also, ofc we cannot have a template for each framework but they can certainly cover the main ones or the ones that are more requested by the community anyway.

**Reply by vplasencia** [reply]
I get your point. I see trade-offs. 

Yes, it can be confusing for some React devs if we don't use the React features they know, even if the project works fine. Maybe they would also like to improve the boilerplate using React best practices, then following this thinking it would be better to keep React features in the boilerplate.  

As an additional idea, we could have a template with HTML, TS (convert it to JS) and CSS where the Semaphore logic can be easily adapted to any framework.

**Reply by cedoor** [reply]
That sounds good. Doesn't Semaphore already have a template like that? Or there's an issue on the backlog maybe

**Reply by vplasencia** [reply]
The template that Semaphore has, is a Nextjs app using CSS instead of ChakraUI. 

There is an issue to build an app with that: https://github.com/semaphore-protocol/semaphore/issues/334

That app can be modified later to create a template. I think it would be better to use TS instead and convert it to JS because most frameworks and devs use TS now.

**Reply by cedoor** [reply]
That makes sense. #334 could be a CLI template directly! Why not?

**Reply by vplasencia** [reply]
Yes, let's do that. Keep the boilerplate with good practices for React and Nextjs and then create a template with HTML, CSS and TS that can be helpful for any JS framework. 

Should the issue title and description https://github.com/semaphore-protocol/semaphore/issues/334 be modified to use TypeScript instead of JavaScript?

**Reply by cedoor** [reply]
Yes, I think so. TypeScript is definitely much more common today.

**Reply by vplasencia** [reply]
Great! I just updated the issue title and description.

**Reply by cedoor** [reply]
Thank you 🙏🏽
